### ISSUE TITLE
Does ants.apply_transforms use inverse transformation?

### ISSUE THREAD

**Issue by cantaylansari-bch** [ANTsX/ANTs#1275]
Hi all.

I have a quick question regarding the working principle of ants.apply_transforms. From the link (https://github.com/ANTsX/ANTs/wiki/Forward-and-inverse-warps-for-warping-images,-pointsets-and-Jacobians)

*`Internally, deforming an image involves transforming a point set in the opposite direction to the intuitive direction of the warping. The "moving" image is being resampled into the fixed space, and the warps transform a sample point (ie, a voxel in the output image) into the moving space. A point at the center of a voxel in the fixed space is transformed to moving space by the forward warps, an interpolated intensity value is computed, and the result is placed in the voxel in the output image.`

Suppose I have a fixed and a moving image and have the forward transformation from the moving to the fixed image. I call apply_transforms as follows:

registered_to_fixed = ants.apply_transforms(fixed = fixed, moving = moving, transformlist = forward_transform_from_moving_to_fixed)

Is the procedure defined at the top* applied in this command? What I mean is that according to the given explanation, in order to warp a moving image into the fixed space, we first need to transform points from the fixed space to the moving space. Then the transformed point is used to generate interpolated intensity and result is placed in the output image in the fixed space. In order to transform a point from the fixed space to the moving space, we need the inverse transformation, but the function only gets forward transformation as parameter. Does it internally take care of it, in other words, does it calculate the inverse transformation to transform the points from the fixed space to the moving space internally?

Regards.

**Comment by ntustison** [comment]
Just to clarify any potential confusion, during the call to antsApplyTransforms (or its equivalent in ANTsPy) an explicit point set is not actually generated.  The transform defines a correspondence from the fixed image to the moving image.  During the call, the transform-generated vector at each voxel in the fixed image is used to sample the corresponding location in the moving image and interpolated according to the specified interpolator.  These voxel-wise interpolated values are used to generate the warped moving image.   So there's no need to worry about the inverse transformation for this specific operation in the case of images.  However, in the case of _actual point sets_, these issues are definitely a concern and are dealt with using different programs.

**Comment by cantaylansari-bch** [comment]
Thank you so much for the quick reply. Correct me if I am wrong but to the best of my knowledge, forward transform has mapping from moving space to fixed space. You told that "The transform defines a correspondence from the fixed image to the moving image.". Isn't that the inverse transform? 

As far as I know (it could be wrong) (also the following procedure is what I would like to do), applyTransforms find a location in the moving space for each voxel in the fixed space using the inverse transform. Then, it gets a bunch of intensity values around that location (according to the interpolation method) in the moving space and interpolate them. The interpolated intensity value is placed to the corresponding voxel in the fixed space. Is that what applyTransform do? Or I need to code it explicitly using applyTransformsToPoints and resample the warped image myself (which could be suffering for me)?

**Comment by ntustison** [comment]
> Thank you so much for the quick reply. Correct me if I am wrong but to the best of my knowledge, forward transform has mapping from moving space to fixed space. You told that "The transform defines a correspondence from the fixed image to the moving image.". Isn't that the inverse transform?

No, I'm correct.  What I described is the forward transform as defined by convention in ITK/ANTs.  Early in its origins, ITK decided on a "pull-back" (i.e., the moving image is *pulled* to the space of the fixed) vs. a "push-to" resampling framework.

**Comment by cantaylansari-bch** [comment]
I see. Actually, what I want to do is "pull-back" transformation as well. As far as I know, if I use a "push-to" resampling, there could be some missing voxels on the target space or some of the transformations could overflow the borders of the target space. And I do not want to do that.

Then, if I call the following:

registered_to_fixed = ants.apply_transforms(fixed = fixed, moving = moving, transformlist = forward_transform_from_moving_to_fixed)

the moving image is pulled to the space of the fixed image, right? 

Lastly, I would like to express my regards for this wonderful framework. It really elevates my research.

**Comment by ntustison** [comment]
>  there could be some missing voxels on the target space or some of the transformations could overflow the borders of the target space. And I do not want to do that.

This was one of the principal motivations in adopting the pull-back framework.

> Then, if I call the following:
```
registered_to_fixed = ants.apply_transforms(fixed = fixed, moving = moving, transformlist = forward_transform_from_moving_to_fixed)
```
> the moving image is pulled to the space of the fixed image, right?

It depends on how you produced ``forward_transform_from_moving_to_fixed``.  The descriptive variable name is confusing from an ANTs/ITK convention standpoint.  The forward transform, conventionally speaking, goes from the fixed image to the moving image.  For example, if you were to call

```
registration = ants.registration(fixed, moving, ...)
```

The resulting forward transform contained in``registration['fwd_transforms']`` would be a "pulling" transform *pointing* from the fixed image to the moving image which would be used by ``ants.apply_transforms(...)`` to pull the moving image into the space of the fixed image.

**Comment by cantaylansari-bch** [comment]
Oh sorry, I forgot to include how I obtained the forward transform. Yes, I am using exactly the same command in order to get the transform:

registration = ants.registration(fixed, moving, ...)

I think I had been confused about the concept of forward transform. I thought that it somehow warps the moving image into the fixed space by applying a push-to approach to moving image. I thought this could result with some missing points or an overflowing issue. But I understand it now. If I use the following command:

registered_to_fixed = ants.apply_transforms(fixed = fixed, moving = moving, transformlist = registration['fwd_transforms'])

, it pulls the moving image into the space of the fixed image and I do not need to worry about any missing voxels or any overflowing issue.

**Comment by cookpa** [comment]
That's right. If you use the forward warps when the target is the fixed space, and the inverse warps when the target is the moving space, `antsApplyTransforms` will pull interpolated intensities correctly into the target space, and your image will not have missing values.
